File: tts-server/sentence_buffer.py
import re


class SentenceBuffer:
    """Accumulates LLM tokens, emits complete sentences for TTS."""

    SENTENCE_ENDERS = frozenset(".!?")
    ABBREVIATIONS = frozenset({
        "dr.", "mr.", "mrs.", "ms.", "jr.", "sr.", "prof.",
        "e.g.", "i.e.", "vs.", "etc.", "no.", "st.", "ft.",
        "u.s.", "u.k.", "a.m.", "p.m.",
    })
    MIN_LENGTH = 20

    def __init__(self):
        self._buffer = ""

    def add_token(self, token: str) -> list[str]:
        """Add a token and return any complete sentences."""
        self._buffer += token
        return self._extract_sentences()

    def flush(self) -> str | None:
        """Return remaining buffered text (call when LLM is done)."""
        text = self._buffer.strip()
        self._buffer = ""
        return text if text else None

    def _extract_sentences(self) -> list[str]:
        sentences = []
        while True:
            split = self._find_split_point()
            if split is None:
                break
            sentence = self._buffer[:split].strip()
            self._buffer = self._buffer[split:]
            if sentence:
                sentences.append(sentence)
        return sentences

    def _find_split_point(self) -> int | None:
        """Find the index to split at, or None if no complete sentence."""
        buf = self._buffer
        # Look for sentence-ending punctuation followed by whitespace
        for i, ch in enumerate(buf):
            if ch in self.SENTENCE_ENDERS and i + 1 < len(buf) and buf[i + 1] in " \n\t\r":
                candidate = buf[: i + 1].strip()
                if len(candidate) < self.MIN_LENGTH:
                    continue
                if self._is_abbreviation(buf, i):
                    continue
                # Split after the punctuation + following whitespace
                split_at = i + 1
                while split_at < len(buf) and buf[split_at] in " \n\t\r":
                    split_at += 1
                return split_at
        return None

    def _is_abbreviation(self, buf: str, dot_pos: int) -> bool:
        """Check if the period at dot_pos is part of an abbreviation."""
        if buf[dot_pos] != ".":
            return False
        # Extract the word ending at dot_pos
        start = dot_pos
        while start > 0 and buf[start - 1].isalpha():
            start -= 1
        # Include the dot
        word = buf[start : dot_pos + 1].lower()
        return word in self.ABBREVIATIONS

File: web-app/src/lib/speech/sentence-buffer.ts
/**
 * SentenceBuffer — TypeScript port of the Python SentenceBuffer.
 *
 * Accumulates LLM tokens and emits complete sentences for TTS.
 * Ported from: backend/services/sentence_buffer.py
 */

const SENTENCE_ENDERS = new Set(['.', '!', '?', '\u3002', '\uFF01', '\uFF1F'])

const ABBREVIATIONS = new Set([
  'dr.',
  'mr.',
  'mrs.',
  'ms.',
  'jr.',
  'sr.',
  'prof.',
  'e.g.',
  'i.e.',
  'vs.',
  'etc.',
  'no.',
  'st.',
  'ft.',
  'u.s.',
  'u.k.',
  'a.m.',
  'p.m.',
])

const MIN_LENGTH = 20

const WHITESPACE = new Set([' ', '\n', '\t', '\r'])

export class SentenceBuffer {
  private buffer = ''

  /**
   * Add a token and return any complete sentences.
   */
  addToken(token: string): string[] {
    this.buffer += token
    return this.extractSentences()
  }

  /**
   * Return remaining buffered text (call when LLM stream is done).
   * Returns null if the buffer is empty.
   */
  flush(): string | null {
    const text = this.buffer.trim()
    this.buffer = ''
    return text || null
  }

  private extractSentences(): string[] {
    const sentences: string[] = []
    while (true) {
      const split = this.findSplitPoint()
      if (split === null) {
        break
      }
      const sentence = this.buffer.slice(0, split).trim()
      this.buffer = this.buffer.slice(split)
      if (sentence) {
        sentences.push(sentence)
      }
    }
    return sentences
  }

  /**
   * Find the index to split at, or null if no complete sentence.
   */
  private findSplitPoint(): number | null {
    const buf = this.buffer

    // Look for sentence-ending punctuation followed by whitespace
    for (let i = 0; i < buf.length; i++) {
      const ch = buf[i]
      if (
        SENTENCE_ENDERS.has(ch) &&
        i + 1 < buf.length &&
        WHITESPACE.has(buf[i + 1])
      ) {
        const candidate = buf.slice(0, i + 1).trim()
        if (candidate.length < MIN_LENGTH) {
          continue
        }
        if (this.isAbbreviation(buf, i)) {
          continue
        }
        // Split after the punctuation + following whitespace
        let splitAt = i + 1
        while (splitAt < buf.length && WHITESPACE.has(buf[splitAt])) {
          splitAt++
        }
        return splitAt
      }
    }
    return null
  }

  /**
   * Check if the period at dotPos is part of an abbreviation.
   */
  private isAbbreviation(buf: string, dotPos: number): boolean {
    if (buf[dotPos] !== '.') {
      return false
    }
    // Extract the word ending at dotPos
    let start = dotPos
    while (start > 0 && /[a-zA-Z]/.test(buf[start - 1])) {
      start--
    }
    // Include the dot
    const word = buf.slice(start, dotPos + 1).toLowerCase()
    return ABBREVIATIONS.has(word)
  }
}
